File: backend-quarkus/src/main/java/app/nanotes/backend/notes/Cursor.java
package app.nanotes.backend.notes;

import app.nanotes.backend.apperr.ValidationException;
import app.nanotes.backend.db.Timestamps;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.time.Instant;
import java.util.Base64;

/**
 * Encodes the stable sort key (updated_at, id) used to page through the
 * notes feed, per ADR 0007. Opaque to clients; only this class constructs
 * or interprets it.
 */
public final class Cursor {

    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final Base64.Encoder ENCODER = Base64.getUrlEncoder().withoutPadding();
    private static final Base64.Decoder DECODER = Base64.getUrlDecoder();

    public final String updatedAt;
    public final String id;

    private Cursor(String updatedAt, String id) {
        this.updatedAt = updatedAt;
        this.id = id;
    }

    private record Payload(@JsonProperty("u") String u, @JsonProperty("i") String i) {}

    public static String encode(Instant updatedAt, String id) {
        try {
            byte[] bytes = MAPPER.writeValueAsBytes(new Payload(Timestamps.format(updatedAt), id));
            return ENCODER.encodeToString(bytes);
        } catch (Exception e) {
            throw new IllegalStateException("failed to encode cursor", e);
        }
    }

    public static Cursor decode(String s) {
        try {
            byte[] bytes = DECODER.decode(s);
            Payload p = MAPPER.readValue(bytes, Payload.class);
            if (p.u() == null || p.u().isEmpty() || p.i() == null || p.i().isEmpty()) {
                throw new ValidationException("invalid cursor");
            }
            return new Cursor(p.u(), p.i());
        } catch (ValidationException e) {
            throw e;
        } catch (Exception e) {
            throw new ValidationException("invalid cursor: " + e.getMessage());
        }
    }
}

File: backend-quarkus/src/main/kotlin/app/nanotes/backend/notes/Cursor.kt
package app.nanotes.backend.notes

import app.nanotes.backend.apperr.ValidationException
import app.nanotes.backend.db.Timestamps
import java.time.Instant
import java.util.Base64

/**
 * Encodes the stable sort key (updated_at, id) used to page through the
 * notes feed, per ADR 0007. Opaque to clients; only this class constructs
 * or interprets it. Each field is base64url-encoded separately and joined
 * with `.` (itself URL-safe), so the whole cursor is a plain URL-safe
 * string with no extra encoding layer needed.
 */
class Cursor private constructor(val updatedAt: String, val id: String) {

    companion object {
        private val ENCODER: Base64.Encoder = Base64.getUrlEncoder().withoutPadding()
        private val DECODER: Base64.Decoder = Base64.getUrlDecoder()

        fun encode(updatedAt: Instant, id: String): String {
            val u = ENCODER.encodeToString(Timestamps.format(updatedAt).toByteArray(Charsets.UTF_8))
            val i = ENCODER.encodeToString(id.toByteArray(Charsets.UTF_8))
            return "$u.$i"
        }

        fun decode(s: String): Cursor {
            val parts = s.split(".", limit = 2)
            if (parts.size != 2) {
                throw ValidationException("invalid cursor")
            }
            return try {
                val u = String(DECODER.decode(parts[0]), Charsets.UTF_8)
                val i = String(DECODER.decode(parts[1]), Charsets.UTF_8)
                if (u.isEmpty() || i.isEmpty()) throw ValidationException("invalid cursor")
                Cursor(u, i)
            } catch (e: ValidationException) {
                throw e
            } catch (e: IllegalArgumentException) {
                throw ValidationException("invalid cursor: ${e.message}")
            }
        }
    }
}
